tools/update_centreon_storage_logs.php

```diff
@@ -0,0 +1,387 @@
+        $dbPassword = askQuestion("Please enter the root password of database: ", true);
+    }
+
+    try {
+        $dsn = sprintf(
+            "mysql:dbname=centreon_storage;host=%s;port=%d",
+            $dbHost,
+            $dbPort
+        );
+        $db = new \PDO($dsn, $dbUser, $dbPassword);
+        $db->setAttribute(\PDO::MYSQL_ATTR_USE_BUFFERED_QUERY, false);
+    } catch (\PDOException $ex) {
+        throw new Exception("Can not connect to the database");
+    }
+
+    /*
+     * First we check if this update is necessary by checking if the log_id
+     * column is present
+     */
+    $result = $db->query(
+        "SELECT COUNT(*) AS is_present
+        FROM information_schema.COLUMNS
+        WHERE TABLE_SCHEMA = 'centreon_storage'
+        AND TABLE_NAME = 'logs' AND COLUMN_NAME = 'log_id'"
+    );
+    $shouldBeContinue = ($result->fetchAll(\PDO::FETCH_ASSOC))[0]['is_present'] == '1';
+
+    if (! $shouldBeContinue) {
+        throw new Exception("Update already done");
+    }
+
+    // We will process partitions that are not empty
+    $partitions = getNotEmptyPartitions($db);
+
+    if (! isset($shouldDeleteOldData)) {
+        // We ask to user if he want to keep old data otherwise we will delete the old logs table
+        $explanation = <<<TEXT
+Before starting, we inform you that we will create a new centreon_storage.logs table and rename the older.
+Next, we will copy data from the older table to new one\n\n
+TEXT;
+        printf($explanation);
+
+        $shouldDeleteOldData =
+            !askYesOrNoQuestion("Do you want to keep the old logs table ?");
+    }
+    // For consistency consideration, we stop broker before creating the new table
+    if ($isBrokerAlreadyStarted) {
+        $logs("For consistency consideration, we stop broker before creating the new table");
+        if (stopBroker() !== 0) {
+            throw new Exception("Error stopping Broker");
+        }
+    } else {
+        $logs("Broker is not started, no need to stop it");
+    }
+    $currentStep++;
+
+    // We create the new logs table and rename older
+    $logs("We create the new logs table and rename older");
+    $db->query("CREATE TABLE logs_new LIKE logs");
+    $currentStep++;
+
+    // Next we delete the column log_id from the new logs table
+    $logs("Next we delete the column log_id from the new logs table");
+    $db->query("ALTER TABLE logs_new DROP COLUMN log_id");
+    $currentStep++;
+
+    // Finally we rename the current table 'logs' to 'logs_old'
+    $logs("Finally we rename the current table 'logs' to 'logs_old'");
+    $db->query("ALTER TABLE logs RENAME TO logs_old");
+    $currentStep++;
+
+    // And we rename the new table 'logs_new' to 'logs'
+    $logs("And we rename the new table 'logs_new' to 'logs'");
+    $db->query("ALTER TABLE logs_new RENAME TO logs");
+    $currentStep++;
+
+    $globalStep = $currentStep;
+
+    // We start Broker if it was previously started
+    if ($isBrokerAlreadyStarted) {
```

Comment:
I would have expected Centreon Broker to be restarted only at the end of the migration process. Otherwise some data could be inserted in the new table and getting this data back would be difficult if the migration has a failure and is aborted.

Note that this also includes table renaming. *logs* should remain named as such until the new table is 100% ready to receive new data.

Reply:
+1